### ISSUE TITLE
signal: 11, SIGSEGV: invalid memory reference (Firefox)

### ISSUE THREAD

**Issue by vnkmpf** [stevepryde/thirtyfour_sync#5]
Hi,

I'm trying to run Selenium with Selenium Hub in Docker.
I keep getting this error:
```
process didn't exit successfully: `/rust/app/foobar/target/debug/deps/selenium-725d0ff37dd25499` (signal: 11, SIGSEGV: invalid memory reference)
```

My thirtyfour_sync version is 0.22.2.

docker-compose.yml:
```yml
version: "3.7"

services:
  rocket:
    build:
      context: docker
    tty: true
    ports:
      - "8000:8000"
    volumes:
      - ./:/rust/app
    working_dir: /rust/app
    networks:
      - selenium-testing

  selenium_hub:
    image: selenium/hub
    networks:
      - selenium-testing

  selenium_firefox:
      image: selenium/node-firefox-debug
      environment:
        - HUB_PORT_4444_TCP_ADDR=selenium_hub
        - HUB_PORT_4444_TCP_PORT=4444
      depends_on:
        - selenium_hub
      networks:
        - selenium-testing

networks:
  selenium-testing:
    driver: bridge
```

The test code:
```rust
use thirtyfour_sync::prelude::*;

#[test]
fn selenium_test() -> color_eyre::Result<()> {
    color_eyre::install()?;

    let mut caps = DesiredCapabilities::firefox();
    caps.set_headless();

    // this line throws segfault
    let driver = WebDriver::new("http://selenium_hub:4444/wd/hub", &caps);

    // here should be more code, but it segfaults on the line above

    Ok(())
}
```

Log:
```
selenium_firefox_1  | 23:45:52.106 INFO [GridLauncherV3.parse] - Selenium server version: 3.141.59, revision: e82be7d358
selenium_hub_1      | 23:45:52.272 INFO [Hub.start] - Selenium Grid hub is up and running
selenium_hub_1      | 23:45:52.282 INFO [Hub.start] - Nodes should register to http://192.168.16.2:4444/grid/register/
selenium_hub_1      | 23:45:52.283 INFO [Hub.start] - Clients should connect to http://192.168.16.2:4444/wd/hub
selenium_firefox_1  | 23:45:52.407 INFO [GridLauncherV3.lambda$buildLaunchers$7] - Launching a Selenium Grid node on port 5555
selenium_firefox_1  | 2021-03-06 23:45:52.673:INFO::main: Logging initialized @1343ms to org.seleniumhq.jetty9.util.log.StdErrLog
selenium_firefox_1  | 23:45:53.236 INFO [WebDriverServlet.<init>] - Initialising WebDriverServlet
selenium_firefox_1  | 23:45:53.472 INFO [SeleniumServer.boot] - Selenium Server is up and running on port 5555
selenium_firefox_1  | 23:45:53.474 INFO [GridLauncherV3.lambda$buildLaunchers$7] - Selenium Grid node is up and ready to register to the hub
selenium_firefox_1  | 23:45:53.584 INFO [SelfRegisteringRemote$1.run] - Starting auto registration thread. Will try to register every 5000 ms.
selenium_firefox_1  | 23:45:54.948 INFO [SelfRegisteringRemote.registerToHub] - Registering the node to the hub: http://selenium_hub:4444/grid/register
selenium_hub_1      | 23:45:55.067 INFO [DefaultGridRegistry.add] - Registered a node http://192.168.16.4:5555
selenium_firefox_1  | 23:45:55.072 INFO [SelfRegisteringRemote.registerToHub] - The node is registered to the hub and ready to use
```

The different IP's and URL in test (selenium_hub) should not be a problem, as I can reach the hub with curl and also when I use the exact same selenium docker images with PHP, it runs without any problems.
I looked at issues in 34-async, nothing helped. I also tried Firefox Standalone image - didn't help.
My Firefox version in the image is 85.0.

Any help is appreciated.

**Comment by stevepryde** [comment]
I'm having trouble replicating your results.

I used the following docker-compose.yml:
```yaml
version: "3"
services:
  selenium_hub:
    image: selenium/hub
    networks:
      - selenium-testing
    ports:
      - "4444:4444"
  selenium_firefox:
    image: selenium/node-firefox-debug
    environment:
      - HUB_PORT_4444_TCP_ADDR=selenium_hub
      - HUB_PORT_4444_TCP_PORT=4444
    depends_on:
      - selenium_hub
    networks:
      - selenium-testing
    ports:
      - "5900:5900"
networks:
  selenium-testing:
    driver: bridge
```

Executable code:
```rust
use thirtyfour_sync::prelude::*;

fn main() -> color_eyre::Result<()> {
    color_eyre::install()?;

    let mut caps = DesiredCapabilities::firefox();
    caps.set_headless()?;
    let driver = WebDriver::new("http://localhost:4444/wd/hub", &caps)?;
    
    Ok(())
}
```
And this works fine for me (thirtyfour_sync = 0.22.2 and firefox = 85.0).

Note I had no success pointing at `http://selenium_hub:4444/wd/hub` as this domain is only known inside docker and the above code is running outside docker. So I exposed the ports in the docker-compose.yml and then `http://localhost:4444/wd/hub` worked. I did not try running the test itself in docker. Let me know if you see the same behaviour both in and outside docker.

Also note the ? after set_headless() and WebDriver::new(). Try adding that in your code - just to see if some error is returned.

What rust version are you using?

**Comment by vnkmpf** [comment]
My Rust version is 1.50.
Another difference then is that I only try it inside docker container (the same as with the PHP I mentioned).
I will try further, esp. outside container, when I get back to the same computer.

**Comment by stevepryde** [comment]
Are you using an alpine base image by any chance? Maybe a dependency lib is missing (e.g. libssl)?. However you seem to be running against rocket and I would have thought rocket requires similar dependencies. Not sure.

**Comment by vnkmpf** [comment]
I got back. I started from the end:
Yes, I do use Rust Alpine image. I did try to install `libressl` and `libressl-dev`, but it did not help.
However, after I tried `rust:1.50.0-buster` image, it works. There's no particular reason why I used Alpine image, so I'm fine with working Debian image.
If someone needs Alpine specifically, I think the solution lies checking the differences between Alpine and Debian images.

Thank you! :)
